# Antens farm

In [31]:
def load_group_data(file, name_blood, name_slh):
    df_blood = pd.read_excel(file, sheet_name=name_blood)
    df_slh = pd.read_excel(file, sheet_name=name_slh)
    df_blood['cow_id'] = df_blood['Land'] + ' ' + df_blood['Levensnr'].astype(str)
    return df_blood, df_slh

In [32]:
def get_same_cows_ids(df_blood, df_slh):
    cow_ids_blood = set(df_blood['cow_id'])
    cow_ids_slh = set(df_slh['ID nummer'])
    return  cow_ids_blood.intersection(cow_ids_slh)

## Group 2348201

In [33]:
import pandas as pd

In [34]:
df_blood, df_slh = load_group_data(
    'Structured data.xlsx', name_blood=0, name_slh=1
)

In [35]:
same_cows_2348201 = get_same_cows_ids(df_blood, df_slh)

In [36]:
len(same_cows_2348201)

549

**There are 549 same cows in both datasets**

## Group 2242208

In [42]:
df_blood_2242208, df_slh_2242208 = load_group_data(
    'Structured data.xlsx',
    name_blood='Group 2242208 - blood samples',
    name_slh='Group 2242208 Slaughterhouse'
)

In [45]:
same_cows_2242208 = get_same_cows_ids(df_blood_2242208, df_slh_2242208)

In [46]:
len(same_cows_2242208)

491

**There are 491 same cows in both datasets**

## Group 2211207 

In [48]:
df_blood_2211207, df_slh_2211207 = load_group_data(
    'Structured data.xlsx',
    name_blood='Group 2211207 - blood samples',
    name_slh='Group 2211207 Slaughterhouse',
)

In [49]:
same_cows_2211207 = get_same_cows_ids(df_blood_2211207, df_slh_2211207)

In [50]:
len(same_cows_2211207)

91

**There are 91 same cows in both datasets**

## Group 2133206

In [51]:
df_blood_2133206, df_slh_2133206 = load_group_data(
    'Structured data.xlsx',
    name_blood='Group 2133206 - blood samples',
    name_slh='Group 2133206 Slaughterhouse',
)

In [52]:
same_cows_2133206 = get_same_cows_ids(df_blood_2133206, df_slh_2133206) 

In [53]:
len(same_cows_2133206)

511

**There are 511 same cows in both datasets**

# Stagri farm

## Group 2338338

In [55]:
df_blood_2338338, df_slh_2338338 = load_group_data(
    'Stagri structured data.xlsx',
    name_blood='Group 2338338 - blood samples',
    name_slh='Group 2338338 Slaughterhouse',
)

In [57]:
same_cows_2338338 = get_same_cows_ids(df_blood_2338338, df_slh_2338338) 

In [58]:
len(same_cows_2338338)

898

## Group 2309337

In [60]:
df_blood_2309337, df_slh_2309337 = load_group_data(
    'Stagri structured data.xlsx',
    name_blood='Group 2309337 - blood samples',
    name_slh='Group 2309337 Slaughterhouse',
)

In [61]:
same_cows_2309337 = get_same_cows_ids(df_blood_2309337, df_slh_2309337) 

In [63]:
len(same_cows_2309337)

152

## Group 2232336

In [64]:
df_blood_2232336, df_slh_2232336 = load_group_data(
    'Stagri structured data.xlsx',
    name_blood='Group 2232336 - blood samples',
    name_slh='Group 2232336 Slaughterhouse',
)

In [65]:
same_cows_2232336 = get_same_cows_ids(df_blood_2232336, df_slh_2232336) 

In [66]:
len(same_cows_2232336)

278

# Group 2128334

In [67]:
df_blood_2128334, df_slh_2128334 = load_group_data(
    'Stagri structured data.xlsx',
    name_blood='Group 2128334 - blood samples',
    name_slh='Group 2128334 Slaughterhouse',
)

In [68]:
same_cows_2128334 = get_same_cows_ids(df_blood_2128334, df_slh_2128334) 

In [69]:
len(same_cows_2128334)

354

## Individual treatments data

In [72]:
treatments = pd.read_excel('Stagri structured data.xlsx', sheet_name='Individual treatments')

In [77]:
treatments.shape

(1484, 15)

In [79]:
treatments['Levensnummer'].unique().shape

(1484,)

In [75]:
cows_in_all_dfs = (same_cows_2338338 | same_cows_2309337 | same_cows_2232336 | same_cows_2128334).intersection(set(treatments['Levensnummer']))

In [76]:
len(cows_in_all_dfs)

898

# Borne farm

## Group 2308249

In [85]:
df_blood_2308249, df_slh_2308249 = load_group_data(
    'Borne structured data.xlsx',
    name_blood='Group 2308249 - blood samples',
    name_slh='Group 2308249 Slaughterhouse',
)

In [86]:
same_cows_2308249 = get_same_cows_ids(df_blood_2308249, df_slh_2308249) 

In [87]:
len(same_cows_2308249)

546

## Group 2229248

In [88]:
df_blood_2229248, df_slh_2229248 = load_group_data(
    'Borne structured data.xlsx',
    name_blood='Group 2229248 - blood samples',
    name_slh='Group 2229248 Slaughterhouse',
)

In [89]:
same_cows_2229248 = get_same_cows_ids(df_blood_2229248, df_slh_2229248) 

In [90]:
len(same_cows_2229248)

303

## Group 2151247

In [91]:
df_blood_2151247, df_slh_2151247 = load_group_data(
    'Borne structured data.xlsx',
    name_blood='Group 2151247 - blood samples',
    name_slh='Group 2151247 Slaughterhouse',
)
same_cows_2151247 = get_same_cows_ids(df_blood_2151247, df_slh_2151247) 
len(same_cows_2151247)

518

## Individual treatments

In [94]:
borne_treatments = pd.read_excel('Borne structured data.xlsx', sheet_name='Individual treatments')

In [99]:
borne_treatments.head(3)

Volgnummer       Datum           Lev.nr  Werknr  Hoknr  \
0 2024-04-03  11/24/2221   DK 03442604712    4712    NaN   
1 2024-03-03   8/14/2024  IE 224075911726   11726    NaN   
2 2024-01-01   8/14/2024  IE 215273412531   12531    NaN   

                 Diagnose             Medicijn Registratienr  Per dier  \
0  Longontsteking pompers  Genta-ject 10% inj.   REG NL 1637         7   
1  Longontsteking pompers          DEPOCILLINE   REG NL 4259        18   
2          kalf staat vol  Buscopan comp. inj.   REG NL 8944         9   

  Dosering Per dier  Naam  Interval            Wachttijd  Wachttijd 2  Dgn  \
0                ml   NaN        24  2222-12-04 00:00:00          NaN  NaN   
1                ml   NaN        24            8/21/2024          NaN  NaN   
2                ml   NaN         0            8/29/2024          NaN  NaN   

   Opmerking  
0        NaN  
1        NaN  
2        NaN

In [96]:
cows_in_all_borne = (same_cows_2308249 | same_cows_2229248 | same_cows_2151247).intersection(set(borne_treatments['Lev.nr']))

In [98]:
len(cows_in_all_borne)

435

## Get sick cows sample

In [101]:
sick_cows_df = borne_treatments[borne_treatments['Lev.nr'].isin(cows_in_all_borne)]
sick_cows_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1307 entries, 1612 to 4978
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Volgnummer         1307 non-null   datetime64[ns]
 1   Datum              1307 non-null   object        
 2   Lev.nr             1307 non-null   object        
 3   Werknr             1307 non-null   int64         
 4   Hoknr              0 non-null      float64       
 5   Diagnose           1307 non-null   object        
 6   Medicijn           1307 non-null   object        
 7   Registratienr      1307 non-null   object        
 8   Per dier           1307 non-null   int64         
 9   Dosering Per dier  1307 non-null   object        
 10  Naam               0 non-null      float64       
 11  Interval           1307 non-null   int64         
 12  Wachttijd          1307 non-null   object        
 13  Wachttijd 2        0 non-null      float64       
 14  Dgn       

In [103]:
sick_cows_df['Datum'] = pd.to_datetime(sick_cows_df['Datum'])

C:\Users\cumpl\AppData\Local\Temp\ipykernel_13476\3366543722.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sick_cows_df['Datum'] = pd.to_datetime(sick_cows_df['Datum'])


In [107]:
sick_cows_df = sick_cows_df.sort_values(by='Datum')
# for now, only keep one record for each sick cow
sick_cows_df = sick_cows_df.drop_duplicates(subset=['Lev.nr'], keep='last')

In [108]:
sick_cows_df

Volgnummer      Datum           Lev.nr  Werknr  Hoknr  \
4950 2024-03-03 2021-12-18    DE 1406617767   17767    NaN   
4943 2024-03-03 2021-12-19    DE 1406576892   76892    NaN   
4931 2024-04-04 2021-12-20    DE 1604501679    1679    NaN   
4885 2024-04-04 2021-12-23     NL 572288820    8882    NaN   
4868 2024-04-04 2021-12-24    DE 1406527581   27581    NaN   
...         ...        ...              ...     ...    ...   
2012 2024-01-01 2023-12-03  IE 216299763534   63534    NaN   
2013 2024-01-01 2023-12-03  IE 215791863069   63069    NaN   
2015 2024-03-03 2023-12-03  IE 215009683161   83161    NaN   
2017 2024-01-01 2023-12-03  IE 224514096375   96375    NaN   
1645 2024-01-01 2023-12-05  IE 214882771709   71709    NaN   

                    Diagnose             Medicijn Registratienr  Per dier  \
4950         Navelontsteking          DEPOCILLINE   REG NL 4259         4   
4943  Longontsteking pompers          DEPOCILLINE   REG NL 4259         4   
4931  Longontsteking pompers  Genta-ject 10% inj.   REG NL 1637         4   
4885  Longontsteking pompers  Genta-ject 10% inj.   REG NL 1637         4   
4868  Longontsteking pompers  Genta-ject 10% inj.   REG NL 1637         5   
...                      ...                  ...           ...       ...   
2012  Longontsteking pompers              Resflor  REG NL 10539         6   
2013  Longontsteking pompers              Resflor  REG NL 10539         6   
2015  Longontsteking pompers     Amoxicilline 100   REG NL 2275         5   
2017  Longontsteking pompers              Resflor  REG NL 10539         6   
1645  Longontsteking pompers              Resflor  REG NL 10539        13   

     Dosering Per dier  Naam  Interval            Wachttijd  Wachttijd 2  Dgn  \
4950                ml   NaN        24            #########          NaN  NaN   
4943                ml   NaN        24            #########          NaN  NaN   
4931                ml   NaN        24  2022-08-05 00:00:00          NaN  NaN   
4885                ml   NaN        24  2022-11-05 00:00:00          NaN  NaN   
4868                ml   NaN        24  2022-12-05 00:00:00          NaN  NaN   
...                ...   ...       ...                  ...          ...  ...   
2012                ml   NaN         0            4/27/2023          NaN  NaN   
2013                ml   NaN         0            4/27/2023          NaN  NaN   
2015                ml   NaN        24  2023-09-04 00:00:00          NaN  NaN   
2017                ml   NaN         0            4/27/2023          NaN  NaN   
1645                ml   NaN         0            6/27/2023          NaN  NaN   

      Opmerking  
4950        NaN  
4943        NaN  
4931        NaN  
4885        NaN  
4868        NaN  
...         ...  
2012        NaN  
2013        NaN  
2015        NaN  
2017        NaN  
1645        NaN  

[435 rows x 16 columns]

# Get non-sick cows sample

In [112]:
non_sick_cows_ids = (same_cows_2308249 | same_cows_2229248 | same_cows_2151247) - set(sick_cows_df['Lev.nr'])

In [113]:
len(non_sick_cows_ids)

932

## Merge datasets

In [136]:
df_blood_2308249.dropna(axis=1, how='all').rename(
    {'cow_id': 'ID nummer'}, axis=1
).set_index('ID nummer').join(
    df_slh_2308249.dropna(axis=1, how='all').set_index('ID nummer'), how='inner'
)

Land      Levensnr  HB1  HB2  HB3  HB4   S2  S3  S4  WerkNr  \
ID nummer                                                                     
IE 222680862571   IE  222680862571  6.4  6.2  NaN  NaN  0.0 NaN NaN   62571   
IE 214158682440   IE  214158682440  5.7  5.1  NaN  NaN  0.0 NaN NaN   82440   
IE 223849832428   IE  223849832428  6.7  5.6  NaN  NaN  0.0 NaN NaN   32428   
IE 217152272253   IE  217152272253  5.2  5.2  NaN  NaN  0.0 NaN NaN   72253   
IE 219154364201   IE  219154364201  6.2  6.4  NaN  NaN  0.0 NaN NaN   64201   
...              ...           ...  ...  ...  ...  ...  ...  ..  ..     ...   
IE 224540571812   IE  224540571812  5.3  5.3  NaN  NaN  0.0 NaN NaN   71812   
IE 219404731629   IE  219404731629  6.4  6.1  NaN  NaN  0.0 NaN NaN   31629   
IE 213733463464   IE  213733463464  5.2  6.8  NaN  NaN  0.0 NaN NaN   63464   
IE 213623280443   IE  213623280443  5.0  5.6  NaN  NaN  0.0 NaN NaN   80443   
IE 217393193773   IE  217393193773  4.7  5.1  NaN  NaN  0.0 NaN NaN   93773   

                    #  Gewicht  Soort  Kleur  Vet        Geboortedatum  \
ID nummer                                                                
IE 222680862571  1308    112.0     13      6    2  2023-01-02 00:00:00   
IE 214158682440  1184    128.9     11      6    2  2023-07-02 00:00:00   
IE 223849832428  1226    108.2     12      7    2  2023-07-02 00:00:00   
IE 217152272253  1209    129.0     11      1    2  2023-07-02 00:00:00   
IE 219154364201  1239     90.9     14      6    1  2023-05-02 00:00:00   
...               ...      ...    ...    ...  ...                  ...   
IE 224540571812   892    158.6     11      6    3            1/28/2023   
IE 219404731629  1314    175.7      9      3    2            1/26/2023   
IE 213733463464   881    135.8     11      5    2            1/17/2023   
IE 213623280443   865    106.1     14      5    1            1/27/2023   
IE 217393193773  1233    118.1     13      7    2            1/27/2023   

                Leeftijdscode   Sekse Slacht Datum  
ID nummer                                           
IE 222680862571             V  STIER     8/23/2023  
IE 214158682440             V  STIER     8/23/2023  
IE 223849832428             V  STIER     8/22/2023  
IE 217152272253             V  STIER     8/22/2023  
IE 219154364201             V  STIER     8/23/2023  
...                       ...     ...          ...  
IE 224540571812             V  STIER     8/22/2023  
IE 219404731629             V  STIER     8/22/2023  
IE 213733463464             V  STIER     8/22/2023  
IE 213623280443             V  STIER     8/22/2023  
IE 217393193773             V  STIER     8/22/2023  

[546 rows x 19 columns]